mlflow/gateway/providers/mosaicml.py

```diff
@@ -0,0 +1,226 @@
+from typing import Any, Dict, List
+
+from fastapi import HTTPException
+from fastapi.encoders import jsonable_encoder
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.config import MosaicMLConfig, RouteConfig
+from mlflow.gateway.providers.base import BaseProvider
+from mlflow.gateway.providers.utils import rename_payload_keys, send_request
+from mlflow.gateway.schemas import chat, completions, embeddings
+
+
+class MosaicMLProvider(BaseProvider):
+    def __init__(self, config: RouteConfig) -> None:
+        super().__init__(config)
+        if config.model.config is None or not isinstance(config.model.config, MosaicMLConfig):
+            raise TypeError(f"Unexpected config type {config.model.config}")
+        self.mosaicml_config: MosaicMLConfig = config.model.config
+
+    async def _request(self, model: str, payload: Dict[str, Any]) -> Dict[str, Any]:
+        headers = {"Authorization": f"{self.mosaicml_config.mosaicml_api_key}"}
+        return await send_request(
+            headers=headers,
+            base_url=self.mosaicml_config.mosaicml_api_base
+            or "https://models.hosted-on.mosaicml.hosting",
+            path=model + "/v1/predict",
+            payload=payload,
+        )
+
+
+    def _parseChatMessagesToPrompt(self, messages: List[chat.RequestMessage]) -> str:
+        """
+        This parser is based on the format described in
+        https://huggingface.co/blog/llama2#how-to-prompt-llama-2 .
+        The expected format is:
+            "<s>[INST] <<SYS>>
+            {{ system_prompt }}
+            <</SYS>>
+
+            {{ user_msg_1 }} [/INST] {{ model_answer_1 }} </s><s>[INST] {{ user_msg_2 }} [/INST]"
+        """
+        if all(m1 != m2 for m1, m2 in zip(messages, messages[1:])):
+            raise MlflowException.invalid_parameter_value(
+                "Consecutive messages cannot have the same 'role'.",
+            )
+        prompt = ""
+        for m in messages:
```

Comment:
We could possibly simplify this by breaking it up.

Add [INST] starter tag.

If there is system or assistant role messags, add the <SYS> wrapper around them. (What do we do if theres' both or multiple?)

Add the user message (hopefully it's present, maybe validation needed :-))  and close off w/ [INST]

Reply:
Or you could do this in the loop, but make sure to pull up adding [INST] as a starter tag outside the loop. We should still clarify what to do w/ duplicated role entries (e.g., multiple system messages -> just concatenate them?).

Reply:
Also, I think <s> </s> separates the entries in a multi turn chat response.

Reply:
I followed the format described here so we should be good: https://huggingface.co/blog/llama2#how-to-prompt-llama-2